tracer/src/Datadog.Trace/Activity/Handlers/DefaultActivityHandler.cs

```diff
@@ -0,0 +1,240 @@
+// <copyright file="DefaultActivityHandler.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Datadog.Trace.Activity.DuckTypes;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Activity.Handlers
+{
+    /// <summary>
+    /// The default handler catches an activity and creates a datadog span from it.
+    /// </summary>
+    internal class DefaultActivityHandler : IActivityHandler
+    {
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(DefaultActivityHandler));
+        private static readonly ConcurrentDictionary<object, Scope> ActivityScope = new();
+
+        public bool ShouldListenTo(string sourceName, string? version)
+        {
+            return true;
+        }
+
+        public void ActivityStarted<T>(string sourceName, T activity)
+            where T : IActivity
+        {
+            var activeSpan = (Span?)Tracer.Instance.ActiveScope?.Span;
+
+            // Propagate Trace and Parent Span ids
+            ulong? traceId = null;
+            ulong? spanId = null;
+            string? rawTraceId = null;
+            string? rawSpanId = null;
+            if (activity is IW3CActivity w3cActivity)
+            {
+                if (activeSpan is not null)
+                {
+                    // If this is the first activity (no parent) and we already have an active span
+                    // or the span was started after the parent activity so we use the span as a parent
+
+                    // We ensure the activity follows the same TraceId as the span
+                    // And marks the ParentId the current spanId
+
+                    if (activity.Parent is null || activity.Parent.StartTimeUtc < activeSpan.StartTime.UtcDateTime)
+                    {
+                        // TraceId
+                        w3cActivity.TraceId = string.IsNullOrWhiteSpace(activeSpan.Context.RawTraceId) ?
+                                                  activeSpan.TraceId.ToString("x32") : activeSpan.Context.RawTraceId;
+
+                        // SpanId
+                        w3cActivity.ParentSpanId = string.IsNullOrWhiteSpace(activeSpan.Context.RawSpanId) ?
+                                                       activeSpan.SpanId.ToString("x16") : activeSpan.Context.RawSpanId;
+
+                        // We clear internals Id and ParentId values to force recalculation.
+                        w3cActivity.RawId = null;
+                        w3cActivity.RawParentId = null;
+
+                        // Avoid recalculation of the traceId.
+                        traceId = activeSpan.TraceId;
```

Comment:
Based on the comment here, maybe you meant to do this?

```suggestion
                        w3cActivity.TraceId = activeSpan.TraceId;
```